### ISSUE TITLE
static keyword

### ISSUE THREAD

**Issue by asymmetric** [dapphub/klab#360]
Rather than using [this heuristic](https://github.com/dapphub/klab/blob/master/lib/build.js#L373) to determine whether to call a function statically, we should make it explicit with a header keyword.

/cc @xwvvvvwx

**Comment by MrChico** [comment]
hhmmmmmmmmmmmmmmmmmm

**Comment by mhhf** [comment]
why?

**Comment by xwvvvvwx** [comment]
the heuristic is kinda magic and hidden. It also breaks for specs that modify storage during their call, but leave storage in the same state by the end of the call (e.g. `transfer-same`).

My concern with adding a `static` keyword is that it would be easy to forget to add it and accidentally leave methods with a looser spec than is strictly required.

**Comment by livnev** [comment]
> the heuristic is kinda magic and hidden. It also breaks for specs that modify storage during their call, but leave storage in the same state by the end of the call (e.g. `transfer-same`).
> 
> My concern with adding a `static` keyword is that it would be easy to forget to add it and accidentally leave methods with a looser spec than is strictly required.

I agree about the heuristic being magic and hidden, I don't like that either really. Your second point about accidentally proving a less general spec is also correct. The way to fix that would be to instead have a `nonstatic` keyword, then if you forget that the worst thing that can happen is that the proof can fail. But this more conservative syntax is a lot of extra overhead... 🤔

**Comment by MrChico** [comment]
I would love to not care about the `static` at all, but unfortunately we need it if we want to be able to compose proofs automatically, since solidity calls some functions with STATICCALL

**Comment by asymmetric** [comment]
Maybe klab should just output whether it thinks the behaviour is static, which would make it more discoverable. Not sure if this would help a new klab user though, and we know where to look already, so don't know if this really helps.
